File: omnigent/server/live_voice.py
"""GPT-Live voice session support: credentials and the WebRTC handshake.

A live voice session is a full-duplex conversation with OpenAI's
``gpt-live-1``, reached over WebRTC. Unlike the spoken summary (one-shot
text in, an MP3 out) this is a *conversation*, so the browser holds the
peer connection and this module only brokers the handshake.

Why the server is in the path at all: the browser must never hold the
OpenAI key. It produces an SDP offer, posts it here, and this module
forwards it with the key attached and hands back the answer. That is the
whole exchange — afterwards audio flows browser-to-OpenAI directly.

**Cost shape, which drives the design.** Sessions bill on wall-clock, not
on speech: silence costs the same as talking. There is no way to *list*
open sessions, so a session whose page vanished is unreachable; ending
one cleanly means sending ``session.close`` over the data channel, which
answers ``session.closed`` with ``reason: close_requested``. Dropping the
peer connection also ends it. Session lifetime is therefore the client's
responsibility, and :data:`MAX_SESSION_S` is a runaway guard rather than
a UX timeout.
"""

from __future__ import annotations

import logging
import os
import pathlib
from typing import Any, Final

import httpx

_logger = logging.getLogger(__name__)

#: OpenAI's live-session endpoint. Creation is WebRTC-only: the POST body
#: carries the caller's SDP offer and the response carries the answer.
LIVE_SESSIONS_URL: Final[str] = "https://api.openai.com/v1/live/sessions"

#: Default conversational model.
DEFAULT_MODEL: Final[str] = "gpt-live-1"

#: Default voice. Lives under ``session.audio.output.voice``; a top-level
#: ``session.voice`` is rejected by the API.
DEFAULT_VOICE: Final[str] = "marin"

#: Where the API key is read from when ``OPENAI_API_KEY`` is unset. Kept
#: outside the repo so it cannot be committed.
KEY_PATH: Final[pathlib.Path] = pathlib.Path.home() / ".omnigent" / "openai-key"

#: Model that reads a finished summary aloud. It decides nothing: the words
#: are already written, and its whole job is to voice them. Pinned small on
#: purpose -- the voice clock runs while the backend thinks, so a slow reader
#: is billed twice, once in its own tokens and again in dead air. Measured:
#: gpt-4o-mini starts speaking in 1.4s, gpt-5 in 27.8s.
DEFAULT_READER_MODEL: Final[str] = "gpt-4o-mini"


def reader_model() -> str:
    """Return the model used to read finished text aloud."""
    return os.environ.get("OMNIGENT_LIVE_READER_MODEL", "").strip() or DEFAULT_READER_MODEL


#: Session prompt for reading, as opposed to conversing.
#:
#: The last clause is load-bearing. A live session generates an opening turn
#: by itself the moment the connection establishes, before any text arrives,
#: and a narrator with nothing to narrate invents something to say -- in
#: testing it fabricated a whole conversation and spoke it confidently.
NARRATOR_INSTRUCTIONS: Final[str] = (
    "You are a narrator with exactly one job: read aloud the text you are given. "
    "Read it naturally, as if telling the listener what happened. Never greet, "
    "never add commentary, never ask a question, never omit anything. "
    "If you have not been given text to read, say absolutely nothing and produce "
    "empty output. Never invent content. "
    "The very first word you speak must be the first word of the text itself. "
    "Never put anything in front of it: no sigh, exhale, hum or throat-clearing, "
    "and no interjection or reaction such as oh, ah, well, so, okay or right."
)


def frame_for_reading(text: str) -> str:
    """Wrap finished prose in the instruction that makes it read, not reply.

    Handed a bare message the reader answers it, which is how a summary
    became unsolicited advice in testing. Told to read it, it reads it.

    :param text: The finished summary, already written by the rewriter.
    :returns: The text framed as a read-aloud command.
    """
    return (
        "Read the following status update aloud to the listener, in your own "
        "speaking voice, keeping every fact and adding nothing. Do not reply to "
        "it, do not advise, do not continue it. Just say it:\n\n"
        f"{text.strip()}"
    )


#: Billed rate, for the client's running-cost estimate. The usage API
#: needs an org-scoped key, so the browser estimates from elapsed time.
USD_PER_MINUTE: Final[float] = 0.05

#: Runaway guard, not a UX timeout: the reader closes sessions by hand.
#: A session that somehow outlives its page still stops billing here.
MAX_SESSION_S: Final[int] = 30 * 60

#: The handshake is one round trip; a slow one means a broken session.
_HANDSHAKE_TIMEOUT_S: Final[float] = 30.0


class LiveVoiceUnavailable(RuntimeError):
    """No usable OpenAI credential, so live voice cannot be offered."""


def api_key() -> str:
    """Return the OpenAI API key.

    :raises LiveVoiceUnavailable: When no key is configured.
    """
    from_env = os.environ.get("OPENAI_API_KEY", "").strip()
    if from_env:
        return from_env
    try:
        from_file = KEY_PATH.read_text(encoding="utf-8").strip()
    except OSError:
        from_file = ""
    if not from_file:
        raise LiveVoiceUnavailable(f"no OpenAI key: set OPENAI_API_KEY or write one to {KEY_PATH}")
    return from_file


def available() -> bool:
    """Whether a live voice session could be created right now."""
    try:
        api_key()
    except LiveVoiceUnavailable:
        return False
    return True


def session_config(
    *,
    instructions: str,
    model: str | None,
    voice: str | None,
    backend: str | None = None,
) -> dict[str, Any]:
    """Build the ``session`` half of a live-session request.

    Field names are load-bearing and were confirmed against the API:
    ``model`` is required, ``instructions`` is accepted, the voice lives
    at ``audio.output.voice``, and unknown keys are rejected outright.

    :param backend: Model to delegate turn content to. ``None`` leaves the
        session in its default ``client`` delegation, where the live model
        answers natively. Naming one switches it to ``responses``
        delegation, the only mode that accepts text pushed in from here.
    """
    config: dict[str, Any] = {
        "model": model or DEFAULT_MODEL,
        "instructions": instructions,
        "audio": {"output": {"voice": voice or DEFAULT_VOICE}},
    }
    if backend:
        config["delegation"] = {"type": "responses", "responses": {"model": backend}}
    return config


async def open_session(
    *,
    sdp_offer: str,
    instructions: str,
    model: str | None = None,
    voice: str | None = None,
    backend: str | None = None,
) -> tuple[str, str]:
    """Exchange a browser SDP offer for a live session.

    :param sdp_offer: The browser's SDP offer.
    :param instructions: System prompt for the conversation.
    :param model: Model override; defaults to :data:`DEFAULT_MODEL`.
    :param voice: Voice override; defaults to :data:`DEFAULT_VOICE`.
    :param backend: Model to delegate turn content to, when text will be
        pushed into the session rather than spoken at it.
    :returns: ``(session_id, sdp_answer)``.
    :raises LiveVoiceUnavailable: When no key is configured.
    :raises httpx.HTTPStatusError: When OpenAI rejects the handshake.
    """
    body = {
        "transport": {"type": "webrtc", "sdp": sdp_offer},
        "session": session_config(
            instructions=instructions, model=model, voice=voice, backend=backend
        ),
    }
    async with httpx.AsyncClient(timeout=_HANDSHAKE_TIMEOUT_S) as client:
        response = await client.post(
            LIVE_SESSIONS_URL,
            headers={"Authorization": f"Bearer {api_key()}"},
            json=body,
        )
    response.raise_for_status()
    payload = response.json()
    session_id = str(payload.get("session", {}).get("id") or "")
    answer = str(payload.get("transport", {}).get("sdp") or "")
    if not answer:
        raise RuntimeError(f"live session returned no SDP answer: {payload!r}")
    _logger.info("live voice session opened: %s", session_id)
    return session_id, answer

File: web/src/lib/liveVoice.ts
/**
 * Speaking a finished summary through OpenAI's live voice model.
 *
 * The words are already written before any of this runs: Gemini wrote them,
 * and nothing here may change them. The live session exists only to say them
 * out loud, which it does in about 1.4 seconds against roughly 23 for local
 * synthesis.
 *
 * The browser holds the peer connection because WebRTC needs a peer, and the
 * server only brokers the handshake so the OpenAI key never reaches here.
 *
 * **The meter runs on wall clock.** A silent session bills exactly as much as
 * a talking one, and the only ways to stop it are `session.close` over the
 * data channel or dropping the connection. So a session is opened once the
 * text exists, never in anticipation, and closed the moment the speech ends.
 */

import { authenticatedFetch } from "./identity";

/** Where the server brokers the SDP exchange. */
const OFFER_URL = "/v1/live/offer";

/**
 * How long to wait after the last spoken word before hanging up.
 *
 * Transcript deltas arrive as the model speaks, so a gap in them means the
 * speech has stopped. Long enough not to clip a pause mid-sentence, short
 * enough that the tail costs a fraction of a cent.
 */
const SILENCE_BEFORE_HANGUP_MS = 2000;

/**
 * How long to wait for the first word before giving up.
 *
 * Much longer than the gap allowed between words: nothing has been said yet,
 * so there is a handshake and a backend turn still to come. Using the short
 * silence timer here hung up mid-thought and the session was billed for a
 * summary nobody heard.
 */
const FIRST_WORD_TIMEOUT_MS = 15_000;

/** Give up if the handshake or the first word never arrives. */
const CONNECT_TIMEOUT_MS = 20_000;

/**
 * How long a conversation may sit silent before it hangs itself up.
 *
 * A conversation bills by wall clock whether anyone is talking or not, and
 * the expensive mistake is a session left open after the reader walked away.
 * Ten seconds of nobody speaking is the signal. Being wrong costs one press
 * of the mic button; not doing it costs dollars an hour against silence.
 */
const CONVERSATION_IDLE_MS = 10_000;

/**
 * Nothing should ever bill longer than this for one summary, whatever goes
 * wrong. A minute of speech is already far past any summary we generate.
 */
const MAX_NARRATION_MS = 90_000;

export interface LiveNarration {
  /** The model's voice, for attaching to an audio element. */
  readonly stream: MediaStream;
  /** Resolves when the speech has finished and the session is closed. */
  readonly finished: Promise<void>;
  /** Hang up now. Safe to call repeatedly. */
  readonly stop: () => void;
}

/** Raised when live narration cannot be used, so the caller can fall back. */
export class LiveVoiceUnavailable extends Error {}

interface OfferResponse {
  session_id: string;
  sdp: string;
  speak: string | null;
}

/**
 * Read `text` aloud through a live session.
 *
 * Resolves as soon as audio is flowing, so the caller can attach the stream
 * and let it play; await `finished` to know when it is over.
 *
 * @param text - The finished summary. Spoken as written, never rewritten.
 * @param options.signal - Aborts the narration and closes the session.
 * @throws LiveVoiceUnavailable when no session could be opened.
 */
export async function narrateViaLive(
  text: string,
  options: { signal?: AbortSignal } = {},
): Promise<LiveNarration> {
  if (!text.trim()) throw new LiveVoiceUnavailable("nothing to read");
  if (typeof RTCPeerConnection === "undefined") {
    throw new LiveVoiceUnavailable("this browser has no WebRTC");
  }

  const pc = new RTCPeerConnection();
  let closed = false;
  let hangupTimer: ReturnType<typeof setTimeout> | undefined;
  let ceilingTimer: ReturnType<typeof setTimeout> | undefined;
  let settleFinished: () => void = () => {};
  const finished = new Promise<void>((resolve) => {
    settleFinished = resolve;
  });

  const stop = (): void => {
    if (closed) return;
    closed = true;
    clearTimeout(hangupTimer);
    clearTimeout(ceilingTimer);
    // Ask for a clean close first: the server acknowledges it and stops
    // billing, where a dropped connection relies on it noticing.
    try {
      if (channel.readyState === "open") {
        channel.send(JSON.stringify({ type: "session.close" }));
      }
    } catch {
      // Already gone; closing the connection below is what matters.
    }
    try {
      pc.close();
    } catch {
      // Nothing left to close.
    }
    settleFinished();
  };

  options.signal?.addEventListener("abort", stop, { once: true });

  const channel = pc.createDataChannel("oai-events");
  const inbound = new MediaStream();
  let firstAudio: (() => void) | undefined;
  const audioArrived = new Promise<void>((resolve) => {
    firstAudio = resolve;
  });

  pc.addEventListener("track", (event) => {
    for (const track of event.streams[0]?.getTracks() ?? [event.track]) {
      inbound.addTrack(track);
    }
    firstAudio?.();
  });

  // A live session negotiates audio both ways, and offering receive-only got
  // no audio back at all. So send a track -- but a silent one synthesized
  // here, never the microphone: narration must not trigger a permission
  // prompt, and there is nothing for it to hear anyway.
  const silence = silentTrack();
  if (silence) pc.addTrack(silence);
  else pc.addTransceiver("audio", { direction: "sendrecv" });

  let spoke = false;
  const bumpHangup = (): void => {
    clearTimeout(hangupTimer);
    hangupTimer = setTimeout(stop, spoke ? SILENCE_BEFORE_HANGUP_MS : FIRST_WORD_TIMEOUT_MS);
  };

  channel.addEventListener("message", (event) => {
    let payload: { type?: string };
    try {
      payload = JSON.parse(String(event.data));
    } catch {
      return;
    }
    // Transcript deltas track the spoken timeline, so they stop when the
    // voice does. That gap, not `response.completed`, marks the real end:
    // the text finishes generating well before it finishes being said.
    if (payload.type === "session.output_transcript.delta") {
      spoke = true;
      bumpHangup();
    }
    if (payload.type === "session.closed") stop();
  });

  const offer = await pc.createOffer();
  await pc.setLocalDescription(offer);

  let answer: OfferResponse;
  try {
    const response = await authenticatedFetch(OFFER_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sdp: pc.localDescription?.sdp ?? offer.sdp, mode: "narrate", text }),
      signal: options.signal,
    });
    if (!response.ok) {
      throw new LiveVoiceUnavailable(`live session refused: ${response.status}`);
    }
    answer = (await response.json()) as OfferResponse;
  } catch (error) {
    stop();
    throw error instanceof LiveVoiceUnavailable ? error : new LiveVoiceUnavailable(String(error));
  }

  if (!answer.speak) {
    stop();
    throw new LiveVoiceUnavailable("server returned nothing to read");
  }

  await pc.setRemoteDescription({ type: "answer", sdp: answer.sdp });

  // Push the text as soon as the channel opens. The session generates an
  // opening turn on its own the moment it connects, and a narrator that
  // reaches that point with nothing in hand invents something to say.
  const speak = answer.speak;
  const push = (): void => {
    channel.send(
      JSON.stringify({
        type: "response.item.create",
        item: { type: "message", role: "user", content: [{ type: "input_text", text: speak }] },
      }),
    );
    channel.send(JSON.stringify({ type: "response.create" }));
    bumpHangup();
  };
  if (channel.readyState === "open") push();
  else channel.addEventListener("open", push, { once: true });

  ceilingTimer = setTimeout(stop, MAX_NARRATION_MS);

  const connected = await Promise.race([
    audioArrived.then(() => true),
    new Promise<false>((resolve) => {
      setTimeout(() => resolve(false), CONNECT_TIMEOUT_MS);
    }),
  ]);
  if (!connected) {
    stop();
    throw new LiveVoiceUnavailable("no audio from the live session");
  }

  return { stream: inbound, finished, stop };
}

/**
 * A silent outbound audio track, so the session has something to negotiate
 * against without opening the microphone.
 *
 * Returns null where Web Audio is unavailable; the caller then falls back to
 * a bare transceiver.
 */
function silentTrack(): MediaStreamTrack | null {
  const Ctor =
    typeof AudioContext !== "undefined"
      ? AudioContext
      : (globalThis as { webkitAudioContext?: typeof AudioContext }).webkitAudioContext;
  if (!Ctor) return null;
  try {
    const context = new Ctor();
    const destination = context.createMediaStreamDestination();
    // An oscillator at zero gain: a running source, emitting nothing.
    const oscillator = context.createOscillator();
    const gain = context.createGain();
    gain.gain.value = 0;
    oscillator.connect(gain).connect(destination);
    oscillator.start();
    return destination.stream.getAudioTracks()[0] ?? null;
  } catch {
    return null;
  }
}

/** One finished utterance, from whichever side said it. */
export interface Utterance {
  /** Who spoke: the reader at the microphone, or the model. */
  readonly who: "reader" | "voice";
  readonly text: string;
}

/**
 * How long a transcript may go quiet before it counts as a finished
 * utterance. Deltas arrive as words are said, so a gap is a pause; short
 * enough to record promptly, long enough not to split a sentence in two.
 */
const UTTERANCE_GAP_MS = 1500;

/** A two-way spoken conversation with the live model. */
export interface LiveConversation {
  /** The model's voice, for attaching to an audio element. */
  readonly stream: MediaStream;
  /** Resolves when the session ends, however it ends. */
  readonly closed: Promise<void>;
  /** Seconds of wall clock billed so far. */
  readonly elapsedS: () => number;
  /** Hang up now. Safe to call repeatedly. */
  readonly stop: () => void;
  /**
   * Resolve once the voice is not mid-sentence, so hanging up does not cut it
   * off. With `expectSpeech`, wait for it to start first: an appended result
   * is spoken shortly after it lands, not at once.
   */
  readonly untilQuiet: (expectSpeech?: boolean) => Promise<void>;
  /**
   * Return a delegation's result for the voice to say in its own words
   * (`session.commentary.append`).
   */
  readonly commentary: (delegationId: string, content: string) => void;
  /** Give the voice something to know without saying it (`session.thinking.append`). */
  readonly thinking: (content: string) => void;
}

/** The pause in the voice's words that marks a sentence as finished. */
const QUIET_GAP_MS = 1200;

/** Never hold a hang-up longer than this waiting for the voice to stop. */
const QUIET_MAX_MS = 4000;

/** How long to wait for the voice to start saying an appended result. */
const QUIET_FIRST_WORD_MS = 3000;

/** The hang-up ceiling when the voice still has something to say. */
const QUIET_EXPECT_MAX_MS = 10_000;

/** How much of the reader's recent speech a delegation carries as its question. */
const MAX_DELEGATED_CHARS = 1000;

/** Each append takes at most 500 tokens; characters are a safe proxy. */
const MAX_APPEND_CHARS = 1800;

/**
 * Open a two-way conversation: it hears the microphone and answers aloud.
 *
 * Unlike narration this has no natural end, so nothing here closes it on a
 * silence. It bills by wall clock for as long as it is open -- roughly three
 * dollars an hour, whether anyone is talking or not -- so the only thing that
 * ends it is the reader, and the caller is expected to show them the meter.
 *
 * The model answers natively from a briefing built from the session's
 * companion ledger. What that does not cover it delegates: GPT-Live's client
 * delegation raises `session.delegation.created`, the caller answers through
 * `commentary`, and the voice says the result.
 *
 * @param sessionId - Conversation whose ledger briefs the model.
 * @throws LiveVoiceUnavailable when the microphone or the session is refused.
 */
export async function openLiveConversation(
  sessionId: string | null,
  options: {
    onUtterance?: (utterance: Utterance) => void;
    /**
     * The voice delegated. `asked` is what the reader said since the last
     * delegation, since the event itself carries no task text.
     */
    onDelegation?: (delegationId: string, asked: string) => void;
    /** Either side was heard: a transcript delta came in for them. */
    onSpeech?: (who: "reader" | "voice") => void;
  } = {},
): Promise<LiveConversation> {
  if (typeof RTCPeerConnection === "undefined") {
    throw new LiveVoiceUnavailable("this browser has no WebRTC");
  }

  let mic: MediaStream;
  try {
    mic = await navigator.mediaDevices.getUserMedia({ audio: true });
  } catch (error) {
    // Refusing the microphone is a choice, not a fault: report it plainly so
    // the caller can say so rather than looking broken.
    throw new LiveVoiceUnavailable(`microphone unavailable: ${String(error)}`);
  }

  const pc = new RTCPeerConnection();
  const startedAt = Date.now();
  let closed = false;
  let settleClosed: () => void = () => {};
  const closedPromise = new Promise<void>((resolve) => {
    settleClosed = resolve;
  });

  const channel = pc.createDataChannel("oai-events");
  const inbound = new MediaStream();
  let idleTimer: ReturnType<typeof setTimeout> | undefined;

  /** Speech from either side means the conversation is alive. */
  const touch = (): void => {
    clearTimeout(idleTimer);
    idleTimer = setTimeout(stop, CONVERSATION_IDLE_MS);
  };

  // Buffer each side's transcript and flush it on a pause. Nothing else
  // records this conversation: audio goes browser-to-OpenAI directly, so if
  // it is not captured here it is gone the moment it is said.
  const buffers: Record<
    "reader" | "voice",
    { text: string; timer?: ReturnType<typeof setTimeout> }
  > = {
    reader: { text: "" },
    voice: { text: "" },
  };

  const collect = (who: "reader" | "voice", delta: string): void => {
    const buffer = buffers[who];
    buffer.text += delta;
    clearTimeout(buffer.timer);
    buffer.timer = setTimeout(() => {
      const said = buffer.text.trim();
      buffer.text = "";
      if (said) options.onUtterance?.({ who, text: said });
    }, UTTERANCE_GAP_MS);
  };

  const flushAll = (): void => {
    for (const who of ["reader", "voice"] as const) {
      const buffer = buffers[who];
      clearTimeout(buffer.timer);
      const said = buffer.text.trim();
      buffer.text = "";
      if (said) options.onUtterance?.({ who, text: said });
    }
  };

  const stop = (): void => {
    if (closed) return;
    closed = true;
    clearTimeout(idleTimer);
    try {
      if (channel.readyState === "open") {
        channel.send(JSON.stringify({ type: "session.close" }));
      }
    } catch {
      // Already gone; closing the connection below is what stops the meter.
    }
    // Whatever was mid-sentence is still worth recording.
    flushAll();
    // Release the microphone, or the browser keeps showing it as in use.
    for (const track of mic.getTracks()) track.stop();
    try {
      pc.close();
    } catch {
      // Nothing left to close.
    }
    settleClosed();
  };

  pc.addEventListener("track", (event) => {
    for (const track of event.streams[0]?.getTracks() ?? [event.track]) {
      inbound.addTrack(track);
    }
  });
  pc.addEventListener("connectionstatechange", () => {
    if (pc.connectionState === "failed" || pc.connectionState === "closed") stop();
  });
  /** Set while waiting for the voice to finish what it is saying. */
  let voiceSpeaking: (() => void) | undefined;
  // What the reader has said since the voice last delegated. A delegation
  // carries no task text, so this is the question it is about.
  let readerSinceDelegation = "";
  let eventCount = 0;

  channel.addEventListener("message", (event) => {
    let payload: {
      type?: string;
      delta?: string;
      delegation?: { id?: string };
      error?: unknown;
    };
    try {
      payload = JSON.parse(String(event.data));
    } catch {
      return;
    }
    if (payload.type === "session.input_transcript.delta") {
      touch();
      if (payload.delta?.trim()) options.onSpeech?.("reader");
      readerSinceDelegation = (readerSinceDelegation + (payload.delta ?? "")).slice(
        -MAX_DELEGATED_CHARS,
      );
      collect("reader", payload.delta ?? "");
    } else if (payload.type === "session.output_transcript.delta") {
      touch();
      voiceSpeaking?.();
      if (payload.delta?.trim()) options.onSpeech?.("voice");
      collect("voice", payload.delta ?? "");
    } else if (payload.type === "session.delegation.created" && payload.delegation?.id) {
      // The voice could not answer and handed the question to this application.
      touch();
      const asked = readerSinceDelegation.trim();
      readerSinceDelegation = "";
      options.onDelegation?.(payload.delegation.id, asked);
    } else if (payload.type === "error") {
      console.warn("live session error", payload.error);
    } else if (payload.type === "session.closed") {
      stop();
    }
  });

  for (const track of mic.getTracks()) pc.addTrack(track, mic);
  // The clock starts at connect, so a session nobody ever speaks into hangs
  // up too rather than billing until the reader notices it is open.
  touch();
  const offer = await pc.createOffer();
  await pc.setLocalDescription(offer);

  try {
    const response = await authenticatedFetch(OFFER_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        sdp: pc.localDescription?.sdp ?? offer.sdp,
        mode: "converse",
        session_id: sessionId,
      }),
    });
    if (!response.ok) {
      throw new LiveVoiceUnavailable(`live session refused: ${response.status}`);
    }
    const answer = (await response.json()) as OfferResponse;
    await pc.setRemoteDescription({ type: "answer", sdp: answer.sdp });
  } catch (error) {
    stop();
    throw error instanceof LiveVoiceUnavailable ? error : new LiveVoiceUnavailable(String(error));
  }

  /** Send one context append, the documented way to feed a running session. */
  const append = (type: string, delegationId: string | null, content: string): void => {
    if (closed || channel.readyState !== "open") return;
    eventCount += 1;
    channel.send(
      JSON.stringify({
        type,
        event_id: `omnigent_${eventCount}`,
        delegation_id: delegationId,
        content: content.slice(0, MAX_APPEND_CHARS),
      }),
    );
  };

  const untilQuiet = (expectSpeech = false): Promise<void> =>
    new Promise<void>((resolve) => {
      if (closed) {
        resolve();
        return;
      }
      let gap: ReturnType<typeof setTimeout> | undefined;
      const done = (): void => {
        clearTimeout(gap);
        clearTimeout(ceiling);
        voiceSpeaking = undefined;
        resolve();
      };
      const ceiling = setTimeout(done, expectSpeech ? QUIET_EXPECT_MAX_MS : QUIET_MAX_MS);
      // Each word from the voice pushes the finish back; a pause ends it.
      voiceSpeaking = () => {
        clearTimeout(gap);
        gap = setTimeout(done, QUIET_GAP_MS);
      };
      gap = setTimeout(done, expectSpeech ? QUIET_FIRST_WORD_MS : QUIET_GAP_MS);
      void closedPromise.then(done);
    });

  return {
    stream: inbound,
    closed: closedPromise,
    elapsedS: () => (Date.now() - startedAt) / 1000,
    stop,
    untilQuiet,
    commentary: (delegationId, content) =>
      append("session.commentary.append", delegationId, content),
    thinking: (content) => append("session.thinking.append", null, content),
  };
}
